# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Function
    source_id_text = [[source_vocab_to_int[y] for y in x.split()] for x in source_text.split('\n')]
    target_id_text = [[target_vocab_to_int[y] for y in x.split()] for x in target_text.split('\n')]
    for sen in target_id_text:
        sen.append(target_vocab_to_int['<EOS>'])
                      
    return source_id_text, target_id_text

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.1.0
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [7]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    import tensorflow as tf
    input = tf.placeholder(tf.int32, [None, None], name='input')
    targets = tf.placeholder(tf.int32, [None, None], name='targets')
    learning_rate = tf.placeholder(tf.float32, None, name="learning_rate")
    keep_prob = tf.placeholder(tf.float32, None, name='keep_prob')
    target_sequence_length = tf.placeholder(tf.int32, (None,), name='target_sequence_length')
    max_target_sequence_length = tf.reduce_max(target_sequence_length, name='max_target_len')
    source_sequence_length = tf.placeholder(tf.int32, (None,), name='source_sequence_length')
    
    return input, targets, learning_rate, keep_prob, target_sequence_length, max_target_sequence_length, \
        source_sequence_length
    

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [8]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    import tensorflow as tf
    ending = tf.strided_slice(target_data, [0, 0], [batch_size, -1], [1, 1])
    dec_input = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), ending], 1)

    return dec_input

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [9]:
from imp import reload
reload(tests)
import tensorflow as tf

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # Encoder embedding
    enc_embed_input = tf.contrib.layers.embed_sequence(rnn_inputs, source_vocab_size, encoding_embedding_size)

    # RNN cell
    def make_cell(rnn_size, keep_prob):
        enc_cell = tf.contrib.rnn.LSTMCell(rnn_size,
                                           initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2))
        enc_cell_with_dropout = tf.contrib.rnn.DropoutWrapper(enc_cell, keep_prob)
        return enc_cell_with_dropout

    enc_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size, keep_prob) for _ in range(num_layers)])
    
    enc_output, enc_state = tf.nn.dynamic_rnn(enc_cell, enc_embed_input, sequence_length=source_sequence_length, dtype=tf.float32)
    
    return enc_output, enc_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [10]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    helper = tf.contrib.seq2seq.TrainingHelper(inputs=dec_embed_input, 
                                               sequence_length=target_sequence_length)
    
    decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, helper, encoder_state, output_layer)
    
    outputs, _ = tf.contrib.seq2seq.dynamic_decode(decoder,
                                                         impute_finished=True,
                                                         maximum_iterations=max_summary_length)

    return outputs

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [11]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    start_tokens = tf.tile(tf.constant([start_of_sequence_id],  
                                       dtype=tf.int32),  
                                       [batch_size], 
                                       name='start_tokens')

    helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings, start_tokens, end_of_sequence_id)
    
    decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, helper, encoder_state, output_layer)
    
    final_outputs, _ = tf.contrib.seq2seq.dynamic_decode(decoder, maximum_iterations=max_target_sequence_length, impute_finished=True)

    return final_outputs

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [12]:
from tensorflow.python.layers.core import Dense

def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)

    # RNN cell
    def make_cell(rnn_size, keep_prob):
        dec_cell = tf.contrib.rnn.LSTMCell(rnn_size,
                                           initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2))
        dec_cell_with_dropout = tf.contrib.rnn.DropoutWrapper(dec_cell, keep_prob)
        return dec_cell_with_dropout

    dec_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size, keep_prob) for _ in range(num_layers)])

    output_layer = Dense(target_vocab_size,
                         kernel_initializer=tf.truncated_normal_initializer(mean=0.0, stddev=0.1))

    with tf.variable_scope("decoding") as decoding_scope:
        training = decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length,
                                       max_target_sequence_length, output_layer, keep_prob)
    
        decoding_scope.reuse_variables()
        infer = decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, target_vocab_to_int['<GO>'],
                             target_vocab_to_int['<EOS>'], max_target_sequence_length,
                             target_vocab_size, output_layer, batch_size, keep_prob)

    return training, infer


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Apply embedding to the input data for the encoder.
- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Apply embedding to the target data for the decoder.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [13]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # Pass the input data through the encoder. We'll ignore the encoder output, but use the state
    _, enc_state = encoding_layer(input_data, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   enc_embedding_size)
    
    
    # Prepare the target sequences we'll feed to the decoder in training mode
    dec_input = process_decoder_input(target_data, target_vocab_to_int, batch_size)
    
    # Pass encoder state and decoder inputs to the decoders
    training_decoder_output, inference_decoder_output = decoding_layer(dec_input, enc_state,
                   target_sequence_length, max_target_sentence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, dec_embedding_size)
    
    return training_decoder_output, inference_decoder_output


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

### Build the Graph
Build the graph using the neural network you implemented.

In [18]:
# Number of Epochs
epochs = 15
# Batch Size
batch_size = 512
# RNN Size
rnn_size = 128
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 200
decoding_embedding_size = 200
# Learning Rate
learning_rate = 0.001
# Dropout Keep Probability
keep_probability = 0.5
display_step = 1

In [19]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [20]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [21]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})

            if batch_i % display_step == 0 and batch_i > 0:

                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})
                
                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch    1/269 - Train Accuracy: 0.2329, Validation Accuracy: 0.3096, Loss: 5.6720
Epoch   0 Batch    2/269 - Train Accuracy: 0.2655, Validation Accuracy: 0.3096, Loss: 5.4426
Epoch   0 Batch    3/269 - Train Accuracy: 0.2444, Validation Accuracy: 0.3096, Loss: 5.2853
Epoch   0 Batch    4/269 - Train Accuracy: 0.2317, Validation Accuracy: 0.3096, Loss: 5.1345
Epoch   0 Batch    5/269 - Train Accuracy: 0.2324, Validation Accuracy: 0.3096, Loss: 4.9566
Epoch   0 Batch    6/269 - Train Accuracy: 0.2785, Validation Accuracy: 0.3096, Loss: 4.6361
Epoch   0 Batch    7/269 - Train Accuracy: 0.2783, Validation Accuracy: 0.3111, Loss: 4.4691
Epoch   0 Batch    8/269 - Train Accuracy: 0.2450, Validation Accuracy: 0.3121, Loss: 4.4728
Epoch   0 Batch    9/269 - Train Accuracy: 0.2798, Validation Accuracy: 0.3214, Loss: 4.2605
Epoch   0 Batch   10/269 - Train Accuracy: 0.2464, Validation Accuracy: 0.3231, Loss: 4.3086
Epoch   0 Batch   11/269 - Train Accuracy: 0.3032, Validation Accuracy

Epoch   0 Batch   90/269 - Train Accuracy: 0.4277, Validation Accuracy: 0.4804, Loss: 2.4881
Epoch   0 Batch   91/269 - Train Accuracy: 0.4514, Validation Accuracy: 0.4779, Loss: 2.3376
Epoch   0 Batch   92/269 - Train Accuracy: 0.4668, Validation Accuracy: 0.4835, Loss: 2.3408
Epoch   0 Batch   93/269 - Train Accuracy: 0.4835, Validation Accuracy: 0.4793, Loss: 2.2486
Epoch   0 Batch   94/269 - Train Accuracy: 0.4582, Validation Accuracy: 0.4847, Loss: 2.3327
Epoch   0 Batch   95/269 - Train Accuracy: 0.4594, Validation Accuracy: 0.4852, Loss: 2.3157
Epoch   0 Batch   96/269 - Train Accuracy: 0.4621, Validation Accuracy: 0.4823, Loss: 2.3095
Epoch   0 Batch   97/269 - Train Accuracy: 0.4569, Validation Accuracy: 0.4789, Loss: 2.3021
Epoch   0 Batch   98/269 - Train Accuracy: 0.4575, Validation Accuracy: 0.4762, Loss: 2.2551
Epoch   0 Batch   99/269 - Train Accuracy: 0.4279, Validation Accuracy: 0.4792, Loss: 2.4087
Epoch   0 Batch  100/269 - Train Accuracy: 0.4873, Validation Accuracy

Epoch   0 Batch  179/269 - Train Accuracy: 0.4817, Validation Accuracy: 0.5062, Loss: 1.6702
Epoch   0 Batch  180/269 - Train Accuracy: 0.4851, Validation Accuracy: 0.5156, Loss: 1.6518
Epoch   0 Batch  181/269 - Train Accuracy: 0.4665, Validation Accuracy: 0.4988, Loss: 1.6538
Epoch   0 Batch  182/269 - Train Accuracy: 0.4634, Validation Accuracy: 0.4932, Loss: 1.6564
Epoch   0 Batch  183/269 - Train Accuracy: 0.5374, Validation Accuracy: 0.5018, Loss: 1.4350
Epoch   0 Batch  184/269 - Train Accuracy: 0.4522, Validation Accuracy: 0.5039, Loss: 1.7077
Epoch   0 Batch  185/269 - Train Accuracy: 0.4731, Validation Accuracy: 0.4946, Loss: 1.6059
Epoch   0 Batch  186/269 - Train Accuracy: 0.4240, Validation Accuracy: 0.4856, Loss: 1.6941
Epoch   0 Batch  187/269 - Train Accuracy: 0.4652, Validation Accuracy: 0.4910, Loss: 1.5878
Epoch   0 Batch  188/269 - Train Accuracy: 0.4884, Validation Accuracy: 0.5012, Loss: 1.5657
Epoch   0 Batch  189/269 - Train Accuracy: 0.4609, Validation Accuracy

Epoch   1 Batch    1/269 - Train Accuracy: 0.4362, Validation Accuracy: 0.4972, Loss: 1.2697
Epoch   1 Batch    2/269 - Train Accuracy: 0.4693, Validation Accuracy: 0.5037, Loss: 1.2418
Epoch   1 Batch    3/269 - Train Accuracy: 0.4643, Validation Accuracy: 0.5133, Loss: 1.2672
Epoch   1 Batch    4/269 - Train Accuracy: 0.4544, Validation Accuracy: 0.5122, Loss: 1.2624
Epoch   1 Batch    5/269 - Train Accuracy: 0.4522, Validation Accuracy: 0.5018, Loss: 1.2727
Epoch   1 Batch    6/269 - Train Accuracy: 0.4761, Validation Accuracy: 0.5007, Loss: 1.1664
Epoch   1 Batch    7/269 - Train Accuracy: 0.4900, Validation Accuracy: 0.5041, Loss: 1.1904
Epoch   1 Batch    8/269 - Train Accuracy: 0.4643, Validation Accuracy: 0.5066, Loss: 1.2538
Epoch   1 Batch    9/269 - Train Accuracy: 0.4652, Validation Accuracy: 0.5044, Loss: 1.2051
Epoch   1 Batch   10/269 - Train Accuracy: 0.4359, Validation Accuracy: 0.4952, Loss: 1.2215
Epoch   1 Batch   11/269 - Train Accuracy: 0.4548, Validation Accuracy

Epoch   1 Batch   90/269 - Train Accuracy: 0.4941, Validation Accuracy: 0.5555, Loss: 0.9975
Epoch   1 Batch   91/269 - Train Accuracy: 0.5233, Validation Accuracy: 0.5559, Loss: 0.9232
Epoch   1 Batch   92/269 - Train Accuracy: 0.5382, Validation Accuracy: 0.5553, Loss: 0.9256
Epoch   1 Batch   93/269 - Train Accuracy: 0.5560, Validation Accuracy: 0.5517, Loss: 0.8952
Epoch   1 Batch   94/269 - Train Accuracy: 0.5398, Validation Accuracy: 0.5600, Loss: 0.9483
Epoch   1 Batch   95/269 - Train Accuracy: 0.5461, Validation Accuracy: 0.5606, Loss: 0.9296
Epoch   1 Batch   96/269 - Train Accuracy: 0.5519, Validation Accuracy: 0.5618, Loss: 0.9209
Epoch   1 Batch   97/269 - Train Accuracy: 0.5313, Validation Accuracy: 0.5537, Loss: 0.9261
Epoch   1 Batch   98/269 - Train Accuracy: 0.5598, Validation Accuracy: 0.5551, Loss: 0.9137
Epoch   1 Batch   99/269 - Train Accuracy: 0.5247, Validation Accuracy: 0.5589, Loss: 0.9625
Epoch   1 Batch  100/269 - Train Accuracy: 0.5672, Validation Accuracy

Epoch   1 Batch  179/269 - Train Accuracy: 0.5767, Validation Accuracy: 0.5829, Loss: 0.8011
Epoch   1 Batch  180/269 - Train Accuracy: 0.5656, Validation Accuracy: 0.5813, Loss: 0.7847
Epoch   1 Batch  181/269 - Train Accuracy: 0.5549, Validation Accuracy: 0.5798, Loss: 0.7952
Epoch   1 Batch  182/269 - Train Accuracy: 0.5738, Validation Accuracy: 0.5818, Loss: 0.7974
Epoch   1 Batch  183/269 - Train Accuracy: 0.6311, Validation Accuracy: 0.5800, Loss: 0.6911
Epoch   1 Batch  184/269 - Train Accuracy: 0.5496, Validation Accuracy: 0.5775, Loss: 0.8161
Epoch   1 Batch  185/269 - Train Accuracy: 0.5790, Validation Accuracy: 0.5739, Loss: 0.7816
Epoch   1 Batch  186/269 - Train Accuracy: 0.5420, Validation Accuracy: 0.5741, Loss: 0.8112
Epoch   1 Batch  187/269 - Train Accuracy: 0.5752, Validation Accuracy: 0.5742, Loss: 0.7759
Epoch   1 Batch  188/269 - Train Accuracy: 0.5865, Validation Accuracy: 0.5746, Loss: 0.7658
Epoch   1 Batch  189/269 - Train Accuracy: 0.5811, Validation Accuracy

Epoch   2 Batch    1/269 - Train Accuracy: 0.5938, Validation Accuracy: 0.6161, Loss: 0.7275
Epoch   2 Batch    2/269 - Train Accuracy: 0.5859, Validation Accuracy: 0.6132, Loss: 0.7108
Epoch   2 Batch    3/269 - Train Accuracy: 0.6002, Validation Accuracy: 0.6111, Loss: 0.7242
Epoch   2 Batch    4/269 - Train Accuracy: 0.5781, Validation Accuracy: 0.6082, Loss: 0.7352
Epoch   2 Batch    5/269 - Train Accuracy: 0.5723, Validation Accuracy: 0.6090, Loss: 0.7348
Epoch   2 Batch    6/269 - Train Accuracy: 0.6024, Validation Accuracy: 0.6110, Loss: 0.6842
Epoch   2 Batch    7/269 - Train Accuracy: 0.6098, Validation Accuracy: 0.6133, Loss: 0.6887
Epoch   2 Batch    8/269 - Train Accuracy: 0.5938, Validation Accuracy: 0.6214, Loss: 0.7369
Epoch   2 Batch    9/269 - Train Accuracy: 0.6031, Validation Accuracy: 0.6197, Loss: 0.7144
Epoch   2 Batch   10/269 - Train Accuracy: 0.5981, Validation Accuracy: 0.6216, Loss: 0.7163
Epoch   2 Batch   11/269 - Train Accuracy: 0.6057, Validation Accuracy

Epoch   2 Batch   90/269 - Train Accuracy: 0.5901, Validation Accuracy: 0.6324, Loss: 0.6801
Epoch   2 Batch   91/269 - Train Accuracy: 0.6347, Validation Accuracy: 0.6309, Loss: 0.6194
Epoch   2 Batch   92/269 - Train Accuracy: 0.6250, Validation Accuracy: 0.6284, Loss: 0.6335
Epoch   2 Batch   93/269 - Train Accuracy: 0.6388, Validation Accuracy: 0.6271, Loss: 0.6104
Epoch   2 Batch   94/269 - Train Accuracy: 0.6292, Validation Accuracy: 0.6245, Loss: 0.6486
Epoch   2 Batch   95/269 - Train Accuracy: 0.6196, Validation Accuracy: 0.6260, Loss: 0.6371
Epoch   2 Batch   96/269 - Train Accuracy: 0.6249, Validation Accuracy: 0.6272, Loss: 0.6420
Epoch   2 Batch   97/269 - Train Accuracy: 0.6121, Validation Accuracy: 0.6273, Loss: 0.6294
Epoch   2 Batch   98/269 - Train Accuracy: 0.6308, Validation Accuracy: 0.6292, Loss: 0.6364
Epoch   2 Batch   99/269 - Train Accuracy: 0.6188, Validation Accuracy: 0.6283, Loss: 0.6537
Epoch   2 Batch  100/269 - Train Accuracy: 0.6456, Validation Accuracy

Epoch   2 Batch  179/269 - Train Accuracy: 0.6365, Validation Accuracy: 0.6373, Loss: 0.5988
Epoch   2 Batch  180/269 - Train Accuracy: 0.6280, Validation Accuracy: 0.6367, Loss: 0.5889
Epoch   2 Batch  181/269 - Train Accuracy: 0.6204, Validation Accuracy: 0.6407, Loss: 0.5920
Epoch   2 Batch  182/269 - Train Accuracy: 0.6495, Validation Accuracy: 0.6388, Loss: 0.5866
Epoch   2 Batch  183/269 - Train Accuracy: 0.6851, Validation Accuracy: 0.6396, Loss: 0.5149
Epoch   2 Batch  184/269 - Train Accuracy: 0.6036, Validation Accuracy: 0.6377, Loss: 0.6086
Epoch   2 Batch  185/269 - Train Accuracy: 0.6416, Validation Accuracy: 0.6375, Loss: 0.5790
Epoch   2 Batch  186/269 - Train Accuracy: 0.6147, Validation Accuracy: 0.6373, Loss: 0.6033
Epoch   2 Batch  187/269 - Train Accuracy: 0.6345, Validation Accuracy: 0.6345, Loss: 0.5767
Epoch   2 Batch  188/269 - Train Accuracy: 0.6404, Validation Accuracy: 0.6352, Loss: 0.5673
Epoch   2 Batch  189/269 - Train Accuracy: 0.6372, Validation Accuracy

Epoch   3 Batch    1/269 - Train Accuracy: 0.6294, Validation Accuracy: 0.6540, Loss: 0.5514
Epoch   3 Batch    2/269 - Train Accuracy: 0.6347, Validation Accuracy: 0.6535, Loss: 0.5353
Epoch   3 Batch    3/269 - Train Accuracy: 0.6434, Validation Accuracy: 0.6526, Loss: 0.5495
Epoch   3 Batch    4/269 - Train Accuracy: 0.6250, Validation Accuracy: 0.6573, Loss: 0.5558
Epoch   3 Batch    5/269 - Train Accuracy: 0.6326, Validation Accuracy: 0.6586, Loss: 0.5554
Epoch   3 Batch    6/269 - Train Accuracy: 0.6583, Validation Accuracy: 0.6616, Loss: 0.5203
Epoch   3 Batch    7/269 - Train Accuracy: 0.6483, Validation Accuracy: 0.6591, Loss: 0.5187
Epoch   3 Batch    8/269 - Train Accuracy: 0.6268, Validation Accuracy: 0.6626, Loss: 0.5610
Epoch   3 Batch    9/269 - Train Accuracy: 0.6475, Validation Accuracy: 0.6639, Loss: 0.5429
Epoch   3 Batch   10/269 - Train Accuracy: 0.6364, Validation Accuracy: 0.6592, Loss: 0.5471
Epoch   3 Batch   11/269 - Train Accuracy: 0.6521, Validation Accuracy

Epoch   3 Batch   90/269 - Train Accuracy: 0.6396, Validation Accuracy: 0.6807, Loss: 0.5126
Epoch   3 Batch   91/269 - Train Accuracy: 0.6891, Validation Accuracy: 0.6862, Loss: 0.4676
Epoch   3 Batch   92/269 - Train Accuracy: 0.6758, Validation Accuracy: 0.6879, Loss: 0.4814
Epoch   3 Batch   93/269 - Train Accuracy: 0.6872, Validation Accuracy: 0.6822, Loss: 0.4598
Epoch   3 Batch   94/269 - Train Accuracy: 0.6818, Validation Accuracy: 0.6818, Loss: 0.4930
Epoch   3 Batch   95/269 - Train Accuracy: 0.6839, Validation Accuracy: 0.6814, Loss: 0.4828
Epoch   3 Batch   96/269 - Train Accuracy: 0.6783, Validation Accuracy: 0.6846, Loss: 0.4831
Epoch   3 Batch   97/269 - Train Accuracy: 0.6854, Validation Accuracy: 0.6836, Loss: 0.4793
Epoch   3 Batch   98/269 - Train Accuracy: 0.6896, Validation Accuracy: 0.6820, Loss: 0.4812
Epoch   3 Batch   99/269 - Train Accuracy: 0.6648, Validation Accuracy: 0.6839, Loss: 0.4956
Epoch   3 Batch  100/269 - Train Accuracy: 0.6941, Validation Accuracy

Epoch   3 Batch  179/269 - Train Accuracy: 0.6950, Validation Accuracy: 0.6996, Loss: 0.4452
Epoch   3 Batch  180/269 - Train Accuracy: 0.6995, Validation Accuracy: 0.7024, Loss: 0.4349
Epoch   3 Batch  181/269 - Train Accuracy: 0.6879, Validation Accuracy: 0.7049, Loss: 0.4444
Epoch   3 Batch  182/269 - Train Accuracy: 0.7143, Validation Accuracy: 0.7119, Loss: 0.4380
Epoch   3 Batch  183/269 - Train Accuracy: 0.7531, Validation Accuracy: 0.7069, Loss: 0.3802
Epoch   3 Batch  184/269 - Train Accuracy: 0.6939, Validation Accuracy: 0.7077, Loss: 0.4525
Epoch   3 Batch  185/269 - Train Accuracy: 0.7269, Validation Accuracy: 0.7135, Loss: 0.4288
Epoch   3 Batch  186/269 - Train Accuracy: 0.7162, Validation Accuracy: 0.7189, Loss: 0.4426
Epoch   3 Batch  187/269 - Train Accuracy: 0.7281, Validation Accuracy: 0.7225, Loss: 0.4303
Epoch   3 Batch  188/269 - Train Accuracy: 0.7246, Validation Accuracy: 0.7211, Loss: 0.4212
Epoch   3 Batch  189/269 - Train Accuracy: 0.7291, Validation Accuracy

Epoch   4 Batch    1/269 - Train Accuracy: 0.7454, Validation Accuracy: 0.7621, Loss: 0.4010
Epoch   4 Batch    2/269 - Train Accuracy: 0.7444, Validation Accuracy: 0.7600, Loss: 0.3879
Epoch   4 Batch    3/269 - Train Accuracy: 0.7503, Validation Accuracy: 0.7587, Loss: 0.4015
Epoch   4 Batch    4/269 - Train Accuracy: 0.7365, Validation Accuracy: 0.7595, Loss: 0.4034
Epoch   4 Batch    5/269 - Train Accuracy: 0.7429, Validation Accuracy: 0.7641, Loss: 0.4031
Epoch   4 Batch    6/269 - Train Accuracy: 0.7807, Validation Accuracy: 0.7598, Loss: 0.3782
Epoch   4 Batch    7/269 - Train Accuracy: 0.7544, Validation Accuracy: 0.7578, Loss: 0.3818
Epoch   4 Batch    8/269 - Train Accuracy: 0.7427, Validation Accuracy: 0.7573, Loss: 0.4090
Epoch   4 Batch    9/269 - Train Accuracy: 0.7680, Validation Accuracy: 0.7543, Loss: 0.3950
Epoch   4 Batch   10/269 - Train Accuracy: 0.7553, Validation Accuracy: 0.7623, Loss: 0.3971
Epoch   4 Batch   11/269 - Train Accuracy: 0.7552, Validation Accuracy

Epoch   4 Batch   90/269 - Train Accuracy: 0.7637, Validation Accuracy: 0.7876, Loss: 0.3630
Epoch   4 Batch   91/269 - Train Accuracy: 0.7987, Validation Accuracy: 0.7854, Loss: 0.3315
Epoch   4 Batch   92/269 - Train Accuracy: 0.7919, Validation Accuracy: 0.7886, Loss: 0.3368
Epoch   4 Batch   93/269 - Train Accuracy: 0.7857, Validation Accuracy: 0.7906, Loss: 0.3243
Epoch   4 Batch   94/269 - Train Accuracy: 0.7819, Validation Accuracy: 0.7926, Loss: 0.3548
Epoch   4 Batch   95/269 - Train Accuracy: 0.7998, Validation Accuracy: 0.7874, Loss: 0.3384
Epoch   4 Batch   96/269 - Train Accuracy: 0.7772, Validation Accuracy: 0.7863, Loss: 0.3416
Epoch   4 Batch   97/269 - Train Accuracy: 0.7976, Validation Accuracy: 0.7905, Loss: 0.3422
Epoch   4 Batch   98/269 - Train Accuracy: 0.7982, Validation Accuracy: 0.7910, Loss: 0.3399
Epoch   4 Batch   99/269 - Train Accuracy: 0.7763, Validation Accuracy: 0.7947, Loss: 0.3506
Epoch   4 Batch  100/269 - Train Accuracy: 0.7956, Validation Accuracy

Epoch   4 Batch  179/269 - Train Accuracy: 0.7862, Validation Accuracy: 0.8144, Loss: 0.3060
Epoch   4 Batch  180/269 - Train Accuracy: 0.8173, Validation Accuracy: 0.8129, Loss: 0.3023
Epoch   4 Batch  181/269 - Train Accuracy: 0.8038, Validation Accuracy: 0.8161, Loss: 0.3099
Epoch   4 Batch  182/269 - Train Accuracy: 0.8116, Validation Accuracy: 0.8154, Loss: 0.3051
Epoch   4 Batch  183/269 - Train Accuracy: 0.8399, Validation Accuracy: 0.8120, Loss: 0.2643
Epoch   4 Batch  184/269 - Train Accuracy: 0.8121, Validation Accuracy: 0.8136, Loss: 0.3060
Epoch   4 Batch  185/269 - Train Accuracy: 0.8156, Validation Accuracy: 0.8113, Loss: 0.2967
Epoch   4 Batch  186/269 - Train Accuracy: 0.8176, Validation Accuracy: 0.8156, Loss: 0.3027
Epoch   4 Batch  187/269 - Train Accuracy: 0.8141, Validation Accuracy: 0.8201, Loss: 0.2947
Epoch   4 Batch  188/269 - Train Accuracy: 0.8182, Validation Accuracy: 0.8211, Loss: 0.2931
Epoch   4 Batch  189/269 - Train Accuracy: 0.8252, Validation Accuracy

Epoch   5 Batch    1/269 - Train Accuracy: 0.8404, Validation Accuracy: 0.8366, Loss: 0.2702
Epoch   5 Batch    2/269 - Train Accuracy: 0.8336, Validation Accuracy: 0.8329, Loss: 0.2650
Epoch   5 Batch    3/269 - Train Accuracy: 0.8341, Validation Accuracy: 0.8362, Loss: 0.2742
Epoch   5 Batch    4/269 - Train Accuracy: 0.8320, Validation Accuracy: 0.8331, Loss: 0.2698
Epoch   5 Batch    5/269 - Train Accuracy: 0.8386, Validation Accuracy: 0.8353, Loss: 0.2732
Epoch   5 Batch    6/269 - Train Accuracy: 0.8557, Validation Accuracy: 0.8346, Loss: 0.2552
Epoch   5 Batch    7/269 - Train Accuracy: 0.8273, Validation Accuracy: 0.8321, Loss: 0.2582
Epoch   5 Batch    8/269 - Train Accuracy: 0.8311, Validation Accuracy: 0.8374, Loss: 0.2747
Epoch   5 Batch    9/269 - Train Accuracy: 0.8337, Validation Accuracy: 0.8367, Loss: 0.2692
Epoch   5 Batch   10/269 - Train Accuracy: 0.8434, Validation Accuracy: 0.8359, Loss: 0.2607
Epoch   5 Batch   11/269 - Train Accuracy: 0.8553, Validation Accuracy

Epoch   5 Batch   90/269 - Train Accuracy: 0.8583, Validation Accuracy: 0.8625, Loss: 0.2369
Epoch   5 Batch   91/269 - Train Accuracy: 0.8752, Validation Accuracy: 0.8612, Loss: 0.2149
Epoch   5 Batch   92/269 - Train Accuracy: 0.8713, Validation Accuracy: 0.8566, Loss: 0.2169
Epoch   5 Batch   93/269 - Train Accuracy: 0.8700, Validation Accuracy: 0.8586, Loss: 0.2148
Epoch   5 Batch   94/269 - Train Accuracy: 0.8558, Validation Accuracy: 0.8591, Loss: 0.2327
Epoch   5 Batch   95/269 - Train Accuracy: 0.8700, Validation Accuracy: 0.8647, Loss: 0.2204
Epoch   5 Batch   96/269 - Train Accuracy: 0.8461, Validation Accuracy: 0.8651, Loss: 0.2278
Epoch   5 Batch   97/269 - Train Accuracy: 0.8687, Validation Accuracy: 0.8639, Loss: 0.2249
Epoch   5 Batch   98/269 - Train Accuracy: 0.8671, Validation Accuracy: 0.8629, Loss: 0.2266
Epoch   5 Batch   99/269 - Train Accuracy: 0.8521, Validation Accuracy: 0.8638, Loss: 0.2305
Epoch   5 Batch  100/269 - Train Accuracy: 0.8751, Validation Accuracy

Epoch   5 Batch  179/269 - Train Accuracy: 0.8620, Validation Accuracy: 0.8732, Loss: 0.1976
Epoch   5 Batch  180/269 - Train Accuracy: 0.8891, Validation Accuracy: 0.8736, Loss: 0.1949
Epoch   5 Batch  181/269 - Train Accuracy: 0.8600, Validation Accuracy: 0.8738, Loss: 0.2034
Epoch   5 Batch  182/269 - Train Accuracy: 0.8817, Validation Accuracy: 0.8760, Loss: 0.1976
Epoch   5 Batch  183/269 - Train Accuracy: 0.8853, Validation Accuracy: 0.8685, Loss: 0.1698
Epoch   5 Batch  184/269 - Train Accuracy: 0.8739, Validation Accuracy: 0.8699, Loss: 0.1965
Epoch   5 Batch  185/269 - Train Accuracy: 0.8718, Validation Accuracy: 0.8732, Loss: 0.1921
Epoch   5 Batch  186/269 - Train Accuracy: 0.8739, Validation Accuracy: 0.8763, Loss: 0.1983
Epoch   5 Batch  187/269 - Train Accuracy: 0.8714, Validation Accuracy: 0.8755, Loss: 0.1898
Epoch   5 Batch  188/269 - Train Accuracy: 0.8662, Validation Accuracy: 0.8742, Loss: 0.1905
Epoch   5 Batch  189/269 - Train Accuracy: 0.8751, Validation Accuracy

Epoch   6 Batch    1/269 - Train Accuracy: 0.8895, Validation Accuracy: 0.8849, Loss: 0.1718
Epoch   6 Batch    2/269 - Train Accuracy: 0.8854, Validation Accuracy: 0.8863, Loss: 0.1690
Epoch   6 Batch    3/269 - Train Accuracy: 0.8933, Validation Accuracy: 0.8840, Loss: 0.1789
Epoch   6 Batch    4/269 - Train Accuracy: 0.8778, Validation Accuracy: 0.8826, Loss: 0.1733
Epoch   6 Batch    5/269 - Train Accuracy: 0.8878, Validation Accuracy: 0.8844, Loss: 0.1743
Epoch   6 Batch    6/269 - Train Accuracy: 0.9072, Validation Accuracy: 0.8804, Loss: 0.1590
Epoch   6 Batch    7/269 - Train Accuracy: 0.8830, Validation Accuracy: 0.8832, Loss: 0.1650
Epoch   6 Batch    8/269 - Train Accuracy: 0.8809, Validation Accuracy: 0.8746, Loss: 0.1734
Epoch   6 Batch    9/269 - Train Accuracy: 0.8750, Validation Accuracy: 0.8754, Loss: 0.1750
Epoch   6 Batch   10/269 - Train Accuracy: 0.8812, Validation Accuracy: 0.8772, Loss: 0.1643
Epoch   6 Batch   11/269 - Train Accuracy: 0.8933, Validation Accuracy

Epoch   6 Batch   90/269 - Train Accuracy: 0.8888, Validation Accuracy: 0.8897, Loss: 0.1537
Epoch   6 Batch   91/269 - Train Accuracy: 0.9129, Validation Accuracy: 0.8896, Loss: 0.1437
Epoch   6 Batch   92/269 - Train Accuracy: 0.8972, Validation Accuracy: 0.8904, Loss: 0.1391
Epoch   6 Batch   93/269 - Train Accuracy: 0.8992, Validation Accuracy: 0.8871, Loss: 0.1417
Epoch   6 Batch   94/269 - Train Accuracy: 0.8863, Validation Accuracy: 0.8887, Loss: 0.1605
Epoch   6 Batch   95/269 - Train Accuracy: 0.8932, Validation Accuracy: 0.8839, Loss: 0.1441
Epoch   6 Batch   96/269 - Train Accuracy: 0.8769, Validation Accuracy: 0.8879, Loss: 0.1519
Epoch   6 Batch   97/269 - Train Accuracy: 0.8952, Validation Accuracy: 0.8895, Loss: 0.1472
Epoch   6 Batch   98/269 - Train Accuracy: 0.8962, Validation Accuracy: 0.8904, Loss: 0.1481
Epoch   6 Batch   99/269 - Train Accuracy: 0.8812, Validation Accuracy: 0.8894, Loss: 0.1531
Epoch   6 Batch  100/269 - Train Accuracy: 0.9007, Validation Accuracy

Epoch   6 Batch  179/269 - Train Accuracy: 0.8890, Validation Accuracy: 0.8998, Loss: 0.1326
Epoch   6 Batch  180/269 - Train Accuracy: 0.9081, Validation Accuracy: 0.8996, Loss: 0.1266
Epoch   6 Batch  181/269 - Train Accuracy: 0.8877, Validation Accuracy: 0.9002, Loss: 0.1415
Epoch   6 Batch  182/269 - Train Accuracy: 0.8932, Validation Accuracy: 0.8991, Loss: 0.1330
Epoch   6 Batch  183/269 - Train Accuracy: 0.9084, Validation Accuracy: 0.9019, Loss: 0.1122
Epoch   6 Batch  184/269 - Train Accuracy: 0.8973, Validation Accuracy: 0.8999, Loss: 0.1280
Epoch   6 Batch  185/269 - Train Accuracy: 0.9076, Validation Accuracy: 0.9002, Loss: 0.1288
Epoch   6 Batch  186/269 - Train Accuracy: 0.8975, Validation Accuracy: 0.9023, Loss: 0.1268
Epoch   6 Batch  187/269 - Train Accuracy: 0.8982, Validation Accuracy: 0.9014, Loss: 0.1274
Epoch   6 Batch  188/269 - Train Accuracy: 0.8930, Validation Accuracy: 0.9014, Loss: 0.1265
Epoch   6 Batch  189/269 - Train Accuracy: 0.8939, Validation Accuracy

Epoch   7 Batch    1/269 - Train Accuracy: 0.9073, Validation Accuracy: 0.9023, Loss: 0.1173
Epoch   7 Batch    2/269 - Train Accuracy: 0.8974, Validation Accuracy: 0.9031, Loss: 0.1174
Epoch   7 Batch    3/269 - Train Accuracy: 0.9176, Validation Accuracy: 0.9023, Loss: 0.1218
Epoch   7 Batch    4/269 - Train Accuracy: 0.8942, Validation Accuracy: 0.9015, Loss: 0.1166
Epoch   7 Batch    5/269 - Train Accuracy: 0.8942, Validation Accuracy: 0.8996, Loss: 0.1179
Epoch   7 Batch    6/269 - Train Accuracy: 0.9180, Validation Accuracy: 0.8986, Loss: 0.1099
Epoch   7 Batch    7/269 - Train Accuracy: 0.9001, Validation Accuracy: 0.8996, Loss: 0.1127
Epoch   7 Batch    8/269 - Train Accuracy: 0.9080, Validation Accuracy: 0.9001, Loss: 0.1201
Epoch   7 Batch    9/269 - Train Accuracy: 0.8990, Validation Accuracy: 0.8987, Loss: 0.1203
Epoch   7 Batch   10/269 - Train Accuracy: 0.9051, Validation Accuracy: 0.8995, Loss: 0.1060
Epoch   7 Batch   11/269 - Train Accuracy: 0.9148, Validation Accuracy

Epoch   7 Batch   90/269 - Train Accuracy: 0.9040, Validation Accuracy: 0.9050, Loss: 0.1102
Epoch   7 Batch   91/269 - Train Accuracy: 0.9241, Validation Accuracy: 0.9054, Loss: 0.0982
Epoch   7 Batch   92/269 - Train Accuracy: 0.9116, Validation Accuracy: 0.9057, Loss: 0.0929
Epoch   7 Batch   93/269 - Train Accuracy: 0.9110, Validation Accuracy: 0.9057, Loss: 0.1011
Epoch   7 Batch   94/269 - Train Accuracy: 0.9033, Validation Accuracy: 0.9047, Loss: 0.1152
Epoch   7 Batch   95/269 - Train Accuracy: 0.8969, Validation Accuracy: 0.9023, Loss: 0.1009
Epoch   7 Batch   96/269 - Train Accuracy: 0.8826, Validation Accuracy: 0.9036, Loss: 0.1087
Epoch   7 Batch   97/269 - Train Accuracy: 0.9023, Validation Accuracy: 0.9027, Loss: 0.1072
Epoch   7 Batch   98/269 - Train Accuracy: 0.9145, Validation Accuracy: 0.9010, Loss: 0.1032
Epoch   7 Batch   99/269 - Train Accuracy: 0.8965, Validation Accuracy: 0.9038, Loss: 0.1049
Epoch   7 Batch  100/269 - Train Accuracy: 0.9154, Validation Accuracy

Epoch   7 Batch  179/269 - Train Accuracy: 0.8999, Validation Accuracy: 0.9107, Loss: 0.0976
Epoch   7 Batch  180/269 - Train Accuracy: 0.9183, Validation Accuracy: 0.9111, Loss: 0.0925
Epoch   7 Batch  181/269 - Train Accuracy: 0.9014, Validation Accuracy: 0.9130, Loss: 0.1034
Epoch   7 Batch  182/269 - Train Accuracy: 0.9039, Validation Accuracy: 0.9149, Loss: 0.0977
Epoch   7 Batch  183/269 - Train Accuracy: 0.9193, Validation Accuracy: 0.9134, Loss: 0.0811
Epoch   7 Batch  184/269 - Train Accuracy: 0.9164, Validation Accuracy: 0.9148, Loss: 0.0937
Epoch   7 Batch  185/269 - Train Accuracy: 0.9157, Validation Accuracy: 0.9095, Loss: 0.0966
Epoch   7 Batch  186/269 - Train Accuracy: 0.9108, Validation Accuracy: 0.9086, Loss: 0.0896
Epoch   7 Batch  187/269 - Train Accuracy: 0.9093, Validation Accuracy: 0.9054, Loss: 0.0954
Epoch   7 Batch  188/269 - Train Accuracy: 0.9098, Validation Accuracy: 0.9081, Loss: 0.0925
Epoch   7 Batch  189/269 - Train Accuracy: 0.9106, Validation Accuracy

Epoch   8 Batch    1/269 - Train Accuracy: 0.9200, Validation Accuracy: 0.9159, Loss: 0.0869
Epoch   8 Batch    2/269 - Train Accuracy: 0.9142, Validation Accuracy: 0.9153, Loss: 0.0867
Epoch   8 Batch    3/269 - Train Accuracy: 0.9263, Validation Accuracy: 0.9165, Loss: 0.0941
Epoch   8 Batch    4/269 - Train Accuracy: 0.8994, Validation Accuracy: 0.9130, Loss: 0.0890
Epoch   8 Batch    5/269 - Train Accuracy: 0.9086, Validation Accuracy: 0.9111, Loss: 0.0890
Epoch   8 Batch    6/269 - Train Accuracy: 0.9301, Validation Accuracy: 0.9124, Loss: 0.0820
Epoch   8 Batch    7/269 - Train Accuracy: 0.9129, Validation Accuracy: 0.9145, Loss: 0.0833
Epoch   8 Batch    8/269 - Train Accuracy: 0.9252, Validation Accuracy: 0.9184, Loss: 0.0888
Epoch   8 Batch    9/269 - Train Accuracy: 0.9158, Validation Accuracy: 0.9173, Loss: 0.0930
Epoch   8 Batch   10/269 - Train Accuracy: 0.9142, Validation Accuracy: 0.9134, Loss: 0.0803
Epoch   8 Batch   11/269 - Train Accuracy: 0.9201, Validation Accuracy

Epoch   8 Batch   90/269 - Train Accuracy: 0.9260, Validation Accuracy: 0.9153, Loss: 0.0856
Epoch   8 Batch   91/269 - Train Accuracy: 0.9368, Validation Accuracy: 0.9173, Loss: 0.0772
Epoch   8 Batch   92/269 - Train Accuracy: 0.9188, Validation Accuracy: 0.9155, Loss: 0.0737
Epoch   8 Batch   93/269 - Train Accuracy: 0.9160, Validation Accuracy: 0.9154, Loss: 0.0781
Epoch   8 Batch   94/269 - Train Accuracy: 0.9249, Validation Accuracy: 0.9177, Loss: 0.0931
Epoch   8 Batch   95/269 - Train Accuracy: 0.9067, Validation Accuracy: 0.9187, Loss: 0.0787
Epoch   8 Batch   96/269 - Train Accuracy: 0.8963, Validation Accuracy: 0.9165, Loss: 0.0858
Epoch   8 Batch   97/269 - Train Accuracy: 0.9187, Validation Accuracy: 0.9173, Loss: 0.0813
Epoch   8 Batch   98/269 - Train Accuracy: 0.9270, Validation Accuracy: 0.9192, Loss: 0.0794
Epoch   8 Batch   99/269 - Train Accuracy: 0.9119, Validation Accuracy: 0.9177, Loss: 0.0822
Epoch   8 Batch  100/269 - Train Accuracy: 0.9251, Validation Accuracy

Epoch   8 Batch  179/269 - Train Accuracy: 0.9133, Validation Accuracy: 0.9188, Loss: 0.0788
Epoch   8 Batch  180/269 - Train Accuracy: 0.9317, Validation Accuracy: 0.9198, Loss: 0.0738
Epoch   8 Batch  181/269 - Train Accuracy: 0.9167, Validation Accuracy: 0.9178, Loss: 0.0819
Epoch   8 Batch  182/269 - Train Accuracy: 0.9172, Validation Accuracy: 0.9208, Loss: 0.0774
Epoch   8 Batch  183/269 - Train Accuracy: 0.9249, Validation Accuracy: 0.9232, Loss: 0.0622
Epoch   8 Batch  184/269 - Train Accuracy: 0.9274, Validation Accuracy: 0.9174, Loss: 0.0748
Epoch   8 Batch  185/269 - Train Accuracy: 0.9279, Validation Accuracy: 0.9182, Loss: 0.0781
Epoch   8 Batch  186/269 - Train Accuracy: 0.9242, Validation Accuracy: 0.9183, Loss: 0.0718
Epoch   8 Batch  187/269 - Train Accuracy: 0.9141, Validation Accuracy: 0.9181, Loss: 0.0761
Epoch   8 Batch  188/269 - Train Accuracy: 0.9207, Validation Accuracy: 0.9213, Loss: 0.0747
Epoch   8 Batch  189/269 - Train Accuracy: 0.9214, Validation Accuracy

Epoch   9 Batch    1/269 - Train Accuracy: 0.9328, Validation Accuracy: 0.9273, Loss: 0.0699
Epoch   9 Batch    2/269 - Train Accuracy: 0.9340, Validation Accuracy: 0.9287, Loss: 0.0723
Epoch   9 Batch    3/269 - Train Accuracy: 0.9387, Validation Accuracy: 0.9267, Loss: 0.0775
Epoch   9 Batch    4/269 - Train Accuracy: 0.9110, Validation Accuracy: 0.9253, Loss: 0.0710
Epoch   9 Batch    5/269 - Train Accuracy: 0.9213, Validation Accuracy: 0.9263, Loss: 0.0725
Epoch   9 Batch    6/269 - Train Accuracy: 0.9385, Validation Accuracy: 0.9270, Loss: 0.0642
Epoch   9 Batch    7/269 - Train Accuracy: 0.9316, Validation Accuracy: 0.9248, Loss: 0.0694
Epoch   9 Batch    8/269 - Train Accuracy: 0.9339, Validation Accuracy: 0.9243, Loss: 0.0733
Epoch   9 Batch    9/269 - Train Accuracy: 0.9271, Validation Accuracy: 0.9213, Loss: 0.0741
Epoch   9 Batch   10/269 - Train Accuracy: 0.9313, Validation Accuracy: 0.9223, Loss: 0.0638
Epoch   9 Batch   11/269 - Train Accuracy: 0.9290, Validation Accuracy

Epoch   9 Batch   90/269 - Train Accuracy: 0.9364, Validation Accuracy: 0.9270, Loss: 0.0682
Epoch   9 Batch   91/269 - Train Accuracy: 0.9425, Validation Accuracy: 0.9314, Loss: 0.0601
Epoch   9 Batch   92/269 - Train Accuracy: 0.9302, Validation Accuracy: 0.9313, Loss: 0.0619
Epoch   9 Batch   93/269 - Train Accuracy: 0.9347, Validation Accuracy: 0.9314, Loss: 0.0655
Epoch   9 Batch   94/269 - Train Accuracy: 0.9283, Validation Accuracy: 0.9339, Loss: 0.0777
Epoch   9 Batch   95/269 - Train Accuracy: 0.9259, Validation Accuracy: 0.9346, Loss: 0.0657
Epoch   9 Batch   96/269 - Train Accuracy: 0.9204, Validation Accuracy: 0.9315, Loss: 0.0729
Epoch   9 Batch   97/269 - Train Accuracy: 0.9295, Validation Accuracy: 0.9323, Loss: 0.0685
Epoch   9 Batch   98/269 - Train Accuracy: 0.9340, Validation Accuracy: 0.9303, Loss: 0.0668
Epoch   9 Batch   99/269 - Train Accuracy: 0.9305, Validation Accuracy: 0.9319, Loss: 0.0652
Epoch   9 Batch  100/269 - Train Accuracy: 0.9396, Validation Accuracy

Epoch   9 Batch  179/269 - Train Accuracy: 0.9237, Validation Accuracy: 0.9363, Loss: 0.0665
Epoch   9 Batch  180/269 - Train Accuracy: 0.9455, Validation Accuracy: 0.9363, Loss: 0.0611
Epoch   9 Batch  181/269 - Train Accuracy: 0.9201, Validation Accuracy: 0.9400, Loss: 0.0693
Epoch   9 Batch  182/269 - Train Accuracy: 0.9340, Validation Accuracy: 0.9387, Loss: 0.0622
Epoch   9 Batch  183/269 - Train Accuracy: 0.9412, Validation Accuracy: 0.9378, Loss: 0.0507
Epoch   9 Batch  184/269 - Train Accuracy: 0.9334, Validation Accuracy: 0.9365, Loss: 0.0591
Epoch   9 Batch  185/269 - Train Accuracy: 0.9327, Validation Accuracy: 0.9371, Loss: 0.0640
Epoch   9 Batch  186/269 - Train Accuracy: 0.9250, Validation Accuracy: 0.9379, Loss: 0.0580
Epoch   9 Batch  187/269 - Train Accuracy: 0.9279, Validation Accuracy: 0.9376, Loss: 0.0615
Epoch   9 Batch  188/269 - Train Accuracy: 0.9314, Validation Accuracy: 0.9362, Loss: 0.0601
Epoch   9 Batch  189/269 - Train Accuracy: 0.9356, Validation Accuracy

Epoch  10 Batch    1/269 - Train Accuracy: 0.9384, Validation Accuracy: 0.9365, Loss: 0.0580
Epoch  10 Batch    2/269 - Train Accuracy: 0.9390, Validation Accuracy: 0.9389, Loss: 0.0596
Epoch  10 Batch    3/269 - Train Accuracy: 0.9419, Validation Accuracy: 0.9377, Loss: 0.0639
Epoch  10 Batch    4/269 - Train Accuracy: 0.9180, Validation Accuracy: 0.9360, Loss: 0.0588
Epoch  10 Batch    5/269 - Train Accuracy: 0.9326, Validation Accuracy: 0.9381, Loss: 0.0585
Epoch  10 Batch    6/269 - Train Accuracy: 0.9451, Validation Accuracy: 0.9359, Loss: 0.0575
Epoch  10 Batch    7/269 - Train Accuracy: 0.9432, Validation Accuracy: 0.9380, Loss: 0.0560
Epoch  10 Batch    8/269 - Train Accuracy: 0.9441, Validation Accuracy: 0.9360, Loss: 0.0623
Epoch  10 Batch    9/269 - Train Accuracy: 0.9311, Validation Accuracy: 0.9362, Loss: 0.0630
Epoch  10 Batch   10/269 - Train Accuracy: 0.9353, Validation Accuracy: 0.9359, Loss: 0.0510
Epoch  10 Batch   11/269 - Train Accuracy: 0.9377, Validation Accuracy

Epoch  10 Batch   90/269 - Train Accuracy: 0.9357, Validation Accuracy: 0.9440, Loss: 0.0609
Epoch  10 Batch   91/269 - Train Accuracy: 0.9432, Validation Accuracy: 0.9423, Loss: 0.0552
Epoch  10 Batch   92/269 - Train Accuracy: 0.9368, Validation Accuracy: 0.9419, Loss: 0.0503
Epoch  10 Batch   93/269 - Train Accuracy: 0.9442, Validation Accuracy: 0.9419, Loss: 0.0548
Epoch  10 Batch   94/269 - Train Accuracy: 0.9323, Validation Accuracy: 0.9434, Loss: 0.0682
Epoch  10 Batch   95/269 - Train Accuracy: 0.9387, Validation Accuracy: 0.9455, Loss: 0.0579
Epoch  10 Batch   96/269 - Train Accuracy: 0.9310, Validation Accuracy: 0.9442, Loss: 0.0619
Epoch  10 Batch   97/269 - Train Accuracy: 0.9381, Validation Accuracy: 0.9434, Loss: 0.0602
Epoch  10 Batch   98/269 - Train Accuracy: 0.9465, Validation Accuracy: 0.9475, Loss: 0.0550
Epoch  10 Batch   99/269 - Train Accuracy: 0.9474, Validation Accuracy: 0.9458, Loss: 0.0555
Epoch  10 Batch  100/269 - Train Accuracy: 0.9446, Validation Accuracy

Epoch  10 Batch  179/269 - Train Accuracy: 0.9331, Validation Accuracy: 0.9419, Loss: 0.0547
Epoch  10 Batch  180/269 - Train Accuracy: 0.9508, Validation Accuracy: 0.9405, Loss: 0.0535
Epoch  10 Batch  181/269 - Train Accuracy: 0.9301, Validation Accuracy: 0.9394, Loss: 0.0586
Epoch  10 Batch  182/269 - Train Accuracy: 0.9321, Validation Accuracy: 0.9382, Loss: 0.0557
Epoch  10 Batch  183/269 - Train Accuracy: 0.9480, Validation Accuracy: 0.9383, Loss: 0.0437
Epoch  10 Batch  184/269 - Train Accuracy: 0.9426, Validation Accuracy: 0.9368, Loss: 0.0524
Epoch  10 Batch  185/269 - Train Accuracy: 0.9368, Validation Accuracy: 0.9362, Loss: 0.0550
Epoch  10 Batch  186/269 - Train Accuracy: 0.9322, Validation Accuracy: 0.9401, Loss: 0.0495
Epoch  10 Batch  187/269 - Train Accuracy: 0.9333, Validation Accuracy: 0.9413, Loss: 0.0542
Epoch  10 Batch  188/269 - Train Accuracy: 0.9419, Validation Accuracy: 0.9403, Loss: 0.0514
Epoch  10 Batch  189/269 - Train Accuracy: 0.9467, Validation Accuracy

Epoch  11 Batch    1/269 - Train Accuracy: 0.9475, Validation Accuracy: 0.9434, Loss: 0.0527
Epoch  11 Batch    2/269 - Train Accuracy: 0.9445, Validation Accuracy: 0.9419, Loss: 0.0514
Epoch  11 Batch    3/269 - Train Accuracy: 0.9462, Validation Accuracy: 0.9430, Loss: 0.0561
Epoch  11 Batch    4/269 - Train Accuracy: 0.9350, Validation Accuracy: 0.9431, Loss: 0.0541
Epoch  11 Batch    5/269 - Train Accuracy: 0.9462, Validation Accuracy: 0.9421, Loss: 0.0512
Epoch  11 Batch    6/269 - Train Accuracy: 0.9479, Validation Accuracy: 0.9401, Loss: 0.0470
Epoch  11 Batch    7/269 - Train Accuracy: 0.9404, Validation Accuracy: 0.9379, Loss: 0.0492
Epoch  11 Batch    8/269 - Train Accuracy: 0.9365, Validation Accuracy: 0.9400, Loss: 0.0551
Epoch  11 Batch    9/269 - Train Accuracy: 0.9379, Validation Accuracy: 0.9432, Loss: 0.0535
Epoch  11 Batch   10/269 - Train Accuracy: 0.9434, Validation Accuracy: 0.9447, Loss: 0.0463
Epoch  11 Batch   11/269 - Train Accuracy: 0.9355, Validation Accuracy

Epoch  11 Batch   90/269 - Train Accuracy: 0.9455, Validation Accuracy: 0.9501, Loss: 0.0526
Epoch  11 Batch   91/269 - Train Accuracy: 0.9545, Validation Accuracy: 0.9465, Loss: 0.0469
Epoch  11 Batch   92/269 - Train Accuracy: 0.9536, Validation Accuracy: 0.9466, Loss: 0.0427
Epoch  11 Batch   93/269 - Train Accuracy: 0.9472, Validation Accuracy: 0.9472, Loss: 0.0493
Epoch  11 Batch   94/269 - Train Accuracy: 0.9406, Validation Accuracy: 0.9525, Loss: 0.0604
Epoch  11 Batch   95/269 - Train Accuracy: 0.9480, Validation Accuracy: 0.9536, Loss: 0.0499
Epoch  11 Batch   96/269 - Train Accuracy: 0.9382, Validation Accuracy: 0.9470, Loss: 0.0536
Epoch  11 Batch   97/269 - Train Accuracy: 0.9453, Validation Accuracy: 0.9463, Loss: 0.0528
Epoch  11 Batch   98/269 - Train Accuracy: 0.9559, Validation Accuracy: 0.9466, Loss: 0.0485
Epoch  11 Batch   99/269 - Train Accuracy: 0.9495, Validation Accuracy: 0.9461, Loss: 0.0493
Epoch  11 Batch  100/269 - Train Accuracy: 0.9515, Validation Accuracy

Epoch  11 Batch  179/269 - Train Accuracy: 0.9394, Validation Accuracy: 0.9484, Loss: 0.0505
Epoch  11 Batch  180/269 - Train Accuracy: 0.9530, Validation Accuracy: 0.9450, Loss: 0.0478
Epoch  11 Batch  181/269 - Train Accuracy: 0.9367, Validation Accuracy: 0.9453, Loss: 0.0535
Epoch  11 Batch  182/269 - Train Accuracy: 0.9410, Validation Accuracy: 0.9435, Loss: 0.0489
Epoch  11 Batch  183/269 - Train Accuracy: 0.9558, Validation Accuracy: 0.9458, Loss: 0.0369
Epoch  11 Batch  184/269 - Train Accuracy: 0.9485, Validation Accuracy: 0.9458, Loss: 0.0458
Epoch  11 Batch  185/269 - Train Accuracy: 0.9473, Validation Accuracy: 0.9454, Loss: 0.0479
Epoch  11 Batch  186/269 - Train Accuracy: 0.9448, Validation Accuracy: 0.9443, Loss: 0.0420
Epoch  11 Batch  187/269 - Train Accuracy: 0.9337, Validation Accuracy: 0.9459, Loss: 0.0474
Epoch  11 Batch  188/269 - Train Accuracy: 0.9512, Validation Accuracy: 0.9474, Loss: 0.0466
Epoch  11 Batch  189/269 - Train Accuracy: 0.9516, Validation Accuracy

Epoch  12 Batch    1/269 - Train Accuracy: 0.9531, Validation Accuracy: 0.9541, Loss: 0.0475
Epoch  12 Batch    2/269 - Train Accuracy: 0.9525, Validation Accuracy: 0.9571, Loss: 0.0470
Epoch  12 Batch    3/269 - Train Accuracy: 0.9472, Validation Accuracy: 0.9557, Loss: 0.0452
Epoch  12 Batch    4/269 - Train Accuracy: 0.9334, Validation Accuracy: 0.9557, Loss: 0.0470
Epoch  12 Batch    5/269 - Train Accuracy: 0.9543, Validation Accuracy: 0.9527, Loss: 0.0463
Epoch  12 Batch    6/269 - Train Accuracy: 0.9524, Validation Accuracy: 0.9519, Loss: 0.0441
Epoch  12 Batch    7/269 - Train Accuracy: 0.9475, Validation Accuracy: 0.9519, Loss: 0.0428
Epoch  12 Batch    8/269 - Train Accuracy: 0.9455, Validation Accuracy: 0.9514, Loss: 0.0499
Epoch  12 Batch    9/269 - Train Accuracy: 0.9447, Validation Accuracy: 0.9475, Loss: 0.0504
Epoch  12 Batch   10/269 - Train Accuracy: 0.9455, Validation Accuracy: 0.9476, Loss: 0.0428
Epoch  12 Batch   11/269 - Train Accuracy: 0.9414, Validation Accuracy

Epoch  12 Batch   90/269 - Train Accuracy: 0.9462, Validation Accuracy: 0.9589, Loss: 0.0472
Epoch  12 Batch   91/269 - Train Accuracy: 0.9599, Validation Accuracy: 0.9644, Loss: 0.0419
Epoch  12 Batch   92/269 - Train Accuracy: 0.9600, Validation Accuracy: 0.9602, Loss: 0.0379
Epoch  12 Batch   93/269 - Train Accuracy: 0.9609, Validation Accuracy: 0.9572, Loss: 0.0426
Epoch  12 Batch   94/269 - Train Accuracy: 0.9443, Validation Accuracy: 0.9582, Loss: 0.0532
Epoch  12 Batch   95/269 - Train Accuracy: 0.9487, Validation Accuracy: 0.9582, Loss: 0.0424
Epoch  12 Batch   96/269 - Train Accuracy: 0.9404, Validation Accuracy: 0.9577, Loss: 0.0485
Epoch  12 Batch   97/269 - Train Accuracy: 0.9519, Validation Accuracy: 0.9554, Loss: 0.0487
Epoch  12 Batch   98/269 - Train Accuracy: 0.9582, Validation Accuracy: 0.9530, Loss: 0.0424
Epoch  12 Batch   99/269 - Train Accuracy: 0.9525, Validation Accuracy: 0.9536, Loss: 0.0442
Epoch  12 Batch  100/269 - Train Accuracy: 0.9523, Validation Accuracy

Epoch  12 Batch  179/269 - Train Accuracy: 0.9487, Validation Accuracy: 0.9543, Loss: 0.0436
Epoch  12 Batch  180/269 - Train Accuracy: 0.9570, Validation Accuracy: 0.9536, Loss: 0.0411
Epoch  12 Batch  181/269 - Train Accuracy: 0.9422, Validation Accuracy: 0.9522, Loss: 0.0500
Epoch  12 Batch  182/269 - Train Accuracy: 0.9478, Validation Accuracy: 0.9514, Loss: 0.0436
Epoch  12 Batch  183/269 - Train Accuracy: 0.9600, Validation Accuracy: 0.9528, Loss: 0.0339
Epoch  12 Batch  184/269 - Train Accuracy: 0.9542, Validation Accuracy: 0.9528, Loss: 0.0426
Epoch  12 Batch  185/269 - Train Accuracy: 0.9552, Validation Accuracy: 0.9500, Loss: 0.0431
Epoch  12 Batch  186/269 - Train Accuracy: 0.9506, Validation Accuracy: 0.9508, Loss: 0.0392
Epoch  12 Batch  187/269 - Train Accuracy: 0.9448, Validation Accuracy: 0.9519, Loss: 0.0439
Epoch  12 Batch  188/269 - Train Accuracy: 0.9548, Validation Accuracy: 0.9489, Loss: 0.0410
Epoch  12 Batch  189/269 - Train Accuracy: 0.9545, Validation Accuracy

Epoch  13 Batch    1/269 - Train Accuracy: 0.9631, Validation Accuracy: 0.9595, Loss: 0.0438
Epoch  13 Batch    2/269 - Train Accuracy: 0.9546, Validation Accuracy: 0.9604, Loss: 0.0396
Epoch  13 Batch    3/269 - Train Accuracy: 0.9554, Validation Accuracy: 0.9616, Loss: 0.0433
Epoch  13 Batch    4/269 - Train Accuracy: 0.9442, Validation Accuracy: 0.9607, Loss: 0.0407
Epoch  13 Batch    5/269 - Train Accuracy: 0.9582, Validation Accuracy: 0.9605, Loss: 0.0419
Epoch  13 Batch    6/269 - Train Accuracy: 0.9658, Validation Accuracy: 0.9564, Loss: 0.0404
Epoch  13 Batch    7/269 - Train Accuracy: 0.9581, Validation Accuracy: 0.9576, Loss: 0.0403
Epoch  13 Batch    8/269 - Train Accuracy: 0.9506, Validation Accuracy: 0.9560, Loss: 0.0449
Epoch  13 Batch    9/269 - Train Accuracy: 0.9508, Validation Accuracy: 0.9593, Loss: 0.0432
Epoch  13 Batch   10/269 - Train Accuracy: 0.9502, Validation Accuracy: 0.9590, Loss: 0.0356
Epoch  13 Batch   11/269 - Train Accuracy: 0.9500, Validation Accuracy

Epoch  13 Batch   90/269 - Train Accuracy: 0.9516, Validation Accuracy: 0.9591, Loss: 0.0413
Epoch  13 Batch   91/269 - Train Accuracy: 0.9597, Validation Accuracy: 0.9591, Loss: 0.0385
Epoch  13 Batch   92/269 - Train Accuracy: 0.9658, Validation Accuracy: 0.9590, Loss: 0.0337
Epoch  13 Batch   93/269 - Train Accuracy: 0.9590, Validation Accuracy: 0.9587, Loss: 0.0375
Epoch  13 Batch   94/269 - Train Accuracy: 0.9520, Validation Accuracy: 0.9586, Loss: 0.0468
Epoch  13 Batch   95/269 - Train Accuracy: 0.9547, Validation Accuracy: 0.9589, Loss: 0.0402
Epoch  13 Batch   96/269 - Train Accuracy: 0.9396, Validation Accuracy: 0.9591, Loss: 0.0423
Epoch  13 Batch   97/269 - Train Accuracy: 0.9552, Validation Accuracy: 0.9603, Loss: 0.0431
Epoch  13 Batch   98/269 - Train Accuracy: 0.9605, Validation Accuracy: 0.9595, Loss: 0.0375
Epoch  13 Batch   99/269 - Train Accuracy: 0.9552, Validation Accuracy: 0.9609, Loss: 0.0382
Epoch  13 Batch  100/269 - Train Accuracy: 0.9575, Validation Accuracy

Epoch  13 Batch  179/269 - Train Accuracy: 0.9500, Validation Accuracy: 0.9606, Loss: 0.0403
Epoch  13 Batch  180/269 - Train Accuracy: 0.9637, Validation Accuracy: 0.9606, Loss: 0.0372
Epoch  13 Batch  181/269 - Train Accuracy: 0.9513, Validation Accuracy: 0.9615, Loss: 0.0404
Epoch  13 Batch  182/269 - Train Accuracy: 0.9501, Validation Accuracy: 0.9611, Loss: 0.0384
Epoch  13 Batch  183/269 - Train Accuracy: 0.9651, Validation Accuracy: 0.9613, Loss: 0.0304
Epoch  13 Batch  184/269 - Train Accuracy: 0.9593, Validation Accuracy: 0.9600, Loss: 0.0362
Epoch  13 Batch  185/269 - Train Accuracy: 0.9593, Validation Accuracy: 0.9607, Loss: 0.0393
Epoch  13 Batch  186/269 - Train Accuracy: 0.9612, Validation Accuracy: 0.9603, Loss: 0.0322
Epoch  13 Batch  187/269 - Train Accuracy: 0.9485, Validation Accuracy: 0.9607, Loss: 0.0373
Epoch  13 Batch  188/269 - Train Accuracy: 0.9608, Validation Accuracy: 0.9610, Loss: 0.0353
Epoch  13 Batch  189/269 - Train Accuracy: 0.9654, Validation Accuracy

Epoch  14 Batch    1/269 - Train Accuracy: 0.9580, Validation Accuracy: 0.9635, Loss: 0.0398
Epoch  14 Batch    2/269 - Train Accuracy: 0.9553, Validation Accuracy: 0.9640, Loss: 0.0387
Epoch  14 Batch    3/269 - Train Accuracy: 0.9565, Validation Accuracy: 0.9655, Loss: 0.0388
Epoch  14 Batch    4/269 - Train Accuracy: 0.9479, Validation Accuracy: 0.9670, Loss: 0.0379
Epoch  14 Batch    5/269 - Train Accuracy: 0.9602, Validation Accuracy: 0.9640, Loss: 0.0367
Epoch  14 Batch    6/269 - Train Accuracy: 0.9634, Validation Accuracy: 0.9590, Loss: 0.0361
Epoch  14 Batch    7/269 - Train Accuracy: 0.9618, Validation Accuracy: 0.9584, Loss: 0.0355
Epoch  14 Batch    8/269 - Train Accuracy: 0.9539, Validation Accuracy: 0.9575, Loss: 0.0404
Epoch  14 Batch    9/269 - Train Accuracy: 0.9474, Validation Accuracy: 0.9568, Loss: 0.0388
Epoch  14 Batch   10/269 - Train Accuracy: 0.9494, Validation Accuracy: 0.9604, Loss: 0.0316
Epoch  14 Batch   11/269 - Train Accuracy: 0.9574, Validation Accuracy

Epoch  14 Batch   90/269 - Train Accuracy: 0.9593, Validation Accuracy: 0.9665, Loss: 0.0365
Epoch  14 Batch   91/269 - Train Accuracy: 0.9654, Validation Accuracy: 0.9630, Loss: 0.0334
Epoch  14 Batch   92/269 - Train Accuracy: 0.9702, Validation Accuracy: 0.9629, Loss: 0.0295
Epoch  14 Batch   93/269 - Train Accuracy: 0.9630, Validation Accuracy: 0.9627, Loss: 0.0340
Epoch  14 Batch   94/269 - Train Accuracy: 0.9527, Validation Accuracy: 0.9627, Loss: 0.0436
Epoch  14 Batch   95/269 - Train Accuracy: 0.9638, Validation Accuracy: 0.9655, Loss: 0.0351
Epoch  14 Batch   96/269 - Train Accuracy: 0.9475, Validation Accuracy: 0.9668, Loss: 0.0399
Epoch  14 Batch   97/269 - Train Accuracy: 0.9598, Validation Accuracy: 0.9680, Loss: 0.0392
Epoch  14 Batch   98/269 - Train Accuracy: 0.9648, Validation Accuracy: 0.9656, Loss: 0.0349
Epoch  14 Batch   99/269 - Train Accuracy: 0.9624, Validation Accuracy: 0.9678, Loss: 0.0337
Epoch  14 Batch  100/269 - Train Accuracy: 0.9646, Validation Accuracy

Epoch  14 Batch  179/269 - Train Accuracy: 0.9553, Validation Accuracy: 0.9624, Loss: 0.0341
Epoch  14 Batch  180/269 - Train Accuracy: 0.9701, Validation Accuracy: 0.9614, Loss: 0.0335
Epoch  14 Batch  181/269 - Train Accuracy: 0.9562, Validation Accuracy: 0.9657, Loss: 0.0377
Epoch  14 Batch  182/269 - Train Accuracy: 0.9566, Validation Accuracy: 0.9652, Loss: 0.0350
Epoch  14 Batch  183/269 - Train Accuracy: 0.9702, Validation Accuracy: 0.9664, Loss: 0.0262
Epoch  14 Batch  184/269 - Train Accuracy: 0.9627, Validation Accuracy: 0.9662, Loss: 0.0334
Epoch  14 Batch  185/269 - Train Accuracy: 0.9697, Validation Accuracy: 0.9651, Loss: 0.0356
Epoch  14 Batch  186/269 - Train Accuracy: 0.9628, Validation Accuracy: 0.9634, Loss: 0.0298
Epoch  14 Batch  187/269 - Train Accuracy: 0.9562, Validation Accuracy: 0.9642, Loss: 0.0347
Epoch  14 Batch  188/269 - Train Accuracy: 0.9635, Validation Accuracy: 0.9616, Loss: 0.0319
Epoch  14 Batch  189/269 - Train Accuracy: 0.9674, Validation Accuracy

Model Trained and Saved


### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [22]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [23]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [24]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    sentence = str.lower(sentence)
    return [vocab_to_int.get(word, vocab_to_int['<UNK>']) for word in sentence.split()]


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [25]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [192, 184, 25, 18, 106, 135, 78]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [83, 180, 307, 20, 298, 234, 357, 1]
  French Words: il a vu un vieux camion . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.